File: Solutions/PAT/Advanced/1051.c
#include <stdio.h>
#include <stdlib.h>
/* Pop Sequence (25) */

int main() {
  int M, N, K; scanf("%d%d%d", &M, &N, &K);
  int stack[1000], top;

  while (K--) {
    int n = 1; /* push n from 1 to N */
    top = -1;
    for (int i=1; i<=N; i++) {
      int k; scanf("%d", &k);
      if (top == M - 1) continue;
      if (top!=-1 && k==stack[top]) top--;
      else {
        while (k!=n && top<M-1 && n<=N) stack[++top] = n++;
        if (k==n && n<N) n++;
      }
    }
    if (top != -1) puts("NO");
    else puts("YES");
  }
}

File: Solutions/PAT/Advanced/1051.py
#!/usr/bin/env python3

M, N, K = [int(i) for i in input().split()]
for i in range(K):
  l = [int(x) for x in input().split()]
  stack = []
  num = 1
  ans = 'NO'
  while True:
    while len(l) > 0 and len(stack) > 0 and stack[-1] == l[0]:
      x = stack.pop()
      l = l[1:]
      if x == N:
        ans = 'YES'
        break
      if num == N + 1 and len(l) > 0 and len(stack) > 0:
        break
      while num != l[0] and num <= N:
        stack.append(num)
        num += 1
      if len(stack) > M:
        break
  print(ans)
